### PR TITLE
Fix bogus warnings for cuda/11.4 with gcc/8.5

### PR DESCRIPTION
**Pull request by ndellingwood**
Workaround for bogus "missing return statement at end of non-void function" warnings

The update applies to possible warnings with cuda version >= 11.3.0 , and does not address occurrences in cuda/11.2

### CODE REVIEW

File: containers/src/Kokkos_DynRankView.hpp [hunk 1]
```diff
@@ -2045,7 +2045,7 @@ inline auto create_mirror_view(
       return Kokkos::Impl::choose_create_mirror(src, arg_prop);
     }
   }
-#if defined KOKKOS_COMPILER_INTEL
+#if defined(KOKKOS_COMPILER_INTEL) || defined(KOKKOS_COMPILER_NVCC)
```
Comment (masterleinad) [inline containers/src/Kokkos_DynRankView.hpp:2048]:
```suggestion
#if defined(KOKKOS_COMPILER_INTEL) || (defined(KOKKOS_COMPILER_NVCC) && KOKKOS_COMPILER_NVCC >= 1130)
```

File: containers/src/Kokkos_DynRankView.hpp [hunk 2]
```diff
@@ -2134,6 +2134,9 @@ auto create_mirror_view_and_copy(
       deep_copy(mirror, src);
     return mirror;
   }
+#if defined(KOKKOS_COMPILER_NVCC)
```
Comment (masterleinad) [inline containers/src/Kokkos_DynRankView.hpp:2137]:
```suggestion
#if defined(KOKKOS_COMPILER_NVCC) && KOKKOS_COMPILER_NVCC >= 1130
```

File: containers/src/Kokkos_DynamicView.hpp [hunk 3]
```diff
@@ -621,7 +621,7 @@ inline auto create_mirror(const Kokkos::Experimental::DynamicView<T, P...>& src,
 
     return ret;
   }
-#if defined KOKKOS_COMPILER_INTEL
+#if defined(KOKKOS_COMPILER_INTEL) || defined(KOKKOS_COMPILER_NVCC)
```
Comment (masterleinad) [inline containers/src/Kokkos_DynamicView.hpp:624]:
```suggestion
#if defined(KOKKOS_COMPILER_INTEL) || (defined(KOKKOS_COMPILER_NVCC) && KOKKOS_COMPILER_NVCC >= 1130)
```

File: containers/src/Kokkos_DynamicView.hpp [hunk 4]
```diff
@@ -716,7 +716,7 @@ inline auto create_mirror_view(
       return Kokkos::Impl::choose_create_mirror(src, arg_prop);
     }
   }
-#if defined KOKKOS_COMPILER_INTEL
+#if defined(KOKKOS_COMPILER_INTEL) || defined(KOKKOS_COMPILER_NVCC)
```
Comment (masterleinad) [inline containers/src/Kokkos_DynamicView.hpp:719]:
```suggestion
#if defined(KOKKOS_COMPILER_INTEL) || (defined(KOKKOS_COMPILER_NVCC) && KOKKOS_COMPILER_NVCC >= 1130)
```

File: containers/src/Kokkos_DynamicView.hpp [hunk 5]
```diff
@@ -998,6 +998,9 @@ auto create_mirror_view_and_copy(
       deep_copy(mirror, src);
     return mirror;
   }
+#if defined(KOKKOS_COMPILER_NVCC)
```
Comment (masterleinad) [inline containers/src/Kokkos_DynamicView.hpp:1001]:
```suggestion
#if defined(KOKKOS_COMPILER_NVCC) && KOKKOS_COMPILER_NVCC >= 1130
```

File: containers/src/Kokkos_OffsetView.hpp [hunk 6]
```diff
@@ -1833,7 +1833,7 @@ inline auto create_mirror(const Kokkos::Experimental::OffsetView<T, P...>& src,
     return typename Kokkos::Experimental::OffsetView<T, P...>::HostMirror(
         Kokkos::create_mirror(arg_prop, src.view()), src.begins());
   }
-#if defined KOKKOS_COMPILER_INTEL
+#if defined(KOKKOS_COMPILER_INTEL) || defined(KOKKOS_COMPILER_NVCC)
```
Comment (masterleinad) [inline containers/src/Kokkos_OffsetView.hpp:1836]:
```suggestion
#if defined(KOKKOS_COMPILER_INTEL) || (defined(KOKKOS_COMPILER_NVCC) && KOKKOS_COMPILER_NVCC >= 1130)
```

File: containers/src/Kokkos_OffsetView.hpp [hunk 7]
```diff
@@ -1927,7 +1927,7 @@ inline auto create_mirror_view(
       return Kokkos::Impl::choose_create_mirror(src, arg_prop);
     }
   }
-#if defined KOKKOS_COMPILER_INTEL
+#if defined(KOKKOS_COMPILER_INTEL) || defined(KOKKOS_COMPILER_NVCC)
```
Comment (masterleinad) [inline containers/src/Kokkos_OffsetView.hpp:1930]:
```suggestion
#if defined(KOKKOS_COMPILER_INTEL) || (defined(KOKKOS_COMPILER_NVCC) && KOKKOS_COMPILER_NVCC >= 1130)
```

File: core/src/Kokkos_CopyViews.hpp [hunk 8]
```diff
@@ -3509,7 +3509,7 @@ inline auto create_mirror(const Kokkos::View<T, P...>& src,
     using dst_type = typename View<T, P...>::HostMirror;
     return dst_type(prop_copy, src.layout());
   }
-#if defined KOKKOS_COMPILER_INTEL
+#if defined(KOKKOS_COMPILER_INTEL) || defined(KOKKOS_COMPILER_NVCC)
```
Comment (masterleinad) [inline core/src/Kokkos_CopyViews.hpp:3512]:
```suggestion
#if defined(KOKKOS_COMPILER_INTEL) || (defined(KOKKOS_COMPILER_NVCC) && KOKKOS_COMPILER_NVCC >= 1130)
```

File: core/src/Kokkos_CopyViews.hpp [hunk 9]
```diff
@@ -3619,7 +3619,7 @@ inline auto choose_create_mirror(
     }
   }
 
-#if defined KOKKOS_COMPILER_INTEL
+#if defined(KOKKOS_COMPILER_INTEL) || defined(KOKKOS_COMPILER_NVCC)
```
Comment (masterleinad) [inline core/src/Kokkos_CopyViews.hpp:3622]:
```suggestion
#if defined(KOKKOS_COMPILER_INTEL) || (defined(KOKKOS_COMPILER_NVCC) && KOKKOS_COMPILER_NVCC >= 1130)
```

File: core/src/Kokkos_CopyViews.hpp [hunk 10]
```diff
@@ -3655,7 +3655,7 @@ inline auto create_mirror_view(
       return Kokkos::Impl::choose_create_mirror(src, arg_prop);
     }
   }
-#if defined KOKKOS_COMPILER_INTEL
+#if defined(KOKKOS_COMPILER_INTEL) || defined(KOKKOS_COMPILER_NVCC)
```
Comment (masterleinad) [inline core/src/Kokkos_CopyViews.hpp:3658]:
```suggestion
#if defined(KOKKOS_COMPILER_INTEL) || (defined(KOKKOS_COMPILER_NVCC) && KOKKOS_COMPILER_NVCC >= 1130)
```

File: core/src/Kokkos_CopyViews.hpp [hunk 11]
```diff
@@ -3763,6 +3763,9 @@ auto create_mirror_view_and_copy(
       deep_copy(mirror, src);
     return mirror;
   }
+#if defined(KOKKOS_COMPILER_NVCC)
```
Comment (masterleinad) [inline core/src/Kokkos_CopyViews.hpp:3766]:
```suggestion
#if defined(KOKKOS_COMPILER_NVCC) && KOKKOS_COMPILER_NVCC >= 1130
```

**Review by masterleinad — CHANGES_REQUESTED** [review]
`__builtin_unreachable` is only available form Cuda 11.3 on.

**Comment by ndellingwood** [comment]
@masterleinad thanks, I committed all the suggestions, ready for re-review

File: containers/src/Kokkos_DynRankView.hpp [hunk 12]
```diff
@@ -1958,7 +1958,7 @@ inline auto create_mirror(const DynRankView<T, P...>& src,
     return dst_type(prop_copy,
                     Impl::reconstructLayout(src.layout(), src.rank()));
   }
-#if defined KOKKOS_COMPILER_INTEL
+#if defined(KOKKOS_COMPILER_INTEL) || defined(KOKKOS_COMPILER_NVCC)
```
Comment (ldh4) [inline containers/src/Kokkos_DynRankView.hpp:1961]: Should `&& defined(KOKKOS_COMPILER_NVCC >= 1130))`  be added here as well?

File: containers/src/Kokkos_OffsetView.hpp [hunk 13]
```diff
@@ -493,7 +493,7 @@ class OffsetView : public ViewTraits<DataType, Properties...> {
       return m_map.m_impl_handle[j0 * m_map.m_impl_offset.m_stride.S0 +
                                  j1 * m_map.m_impl_offset.m_stride.S1];
     }
-#if defined KOKKOS_COMPILER_INTEL
+#if defined(KOKKOS_COMPILER_INTEL)
```
Comment (ldh4) [inline containers/src/Kokkos_OffsetView.hpp:496]: Is `(defined(KOKKOS_COMPILER_NVCC) && KOKKOS_COMPILER_NVCC >= 1130)` not needed for this line?

Comment (ndellingwood) [reply containers/src/Kokkos_OffsetView.hpp:496]: No, this location didn't trigger the warning, and it is within a kernel (I tested with the added guard and it warned about `__builtin_unreachable()` lacking device markings)

File: containers/src/Kokkos_DynRankView.hpp (hunk 12 above)
Comment (ndellingwood) [reply containers/src/Kokkos_DynRankView.hpp]: Thanks, fixed with 7e974b93cc25b43c778e4dd4eb8b8439be0b2ae7

**Review by dalg24 — COMMENTED** [review]
Does this mean we are giving up on CUDA 11.2?

**Comment by masterleinad** [comment]
> Does this mean we are giving up on CUDA 11.2?

I would rather not try to accommodate 11.2 (which isn't even available on most of the big systems anymore) apart from possibly trying to suppress these warnings.

**Comment by dalg24** [comment]
That would be fine by me.  Just pointing out that, if that is what we decide, we need to update the title and clearly state in the description that we are not fixing 11.2

**Comment by dalg24** [comment]
Seeing `D:/a/kokkos/kokkos/core/src\Kokkos_CopyViews.hpp(3661): error C3861: '__builtin_unreachable': identifier not found` in the windows CUDA build

**Comment by masterleinad** [comment]
> Seeing `D:/a/kokkos/kokkos/core/src\Kokkos_CopyViews.hpp(3661): error C3861: '__builtin_unreachable': identifier not found` in the windows CUDA build

We noticed before that this doesn't work with MSVC, see https://github.com/kokkos/kokkos/blob/7c0ccd298b45641c216faecb93d5241201f56159/core/unit_test/TestMathematicalFunctions.hpp#L33-L39.

**Review by dalg24 — APPROVED** [review]

**Comment by ndellingwood** [comment]
Thanks @masterleinad , I'll add approval but not sure it'll count since I submitted the initial PR

**Comment by ndellingwood** [comment]
Nevermind, approval option blocked

**Comment by ndellingwood** [comment]
Cherry-picked to release-candidate-4.4.00, PR https://github.com/kokkos/kokkos/pull/7185

**Pull request merged**